test/test_prototype_datasets.py

```diff
@@ -1,2 +1,250 @@
+            assert f"{key}={value}" in output
+
+
+class TestDatasetInfo:
+    @pytest.fixture
+    def info(self):
+        return make_minimal_dataset_info(valid_options=dict(split=("train", "test"), foo=("bar", "baz")))
+
+    def test_default_config(self, info):
+        valid_options = info._valid_options
+        default_config = datasets.utils.DatasetConfig({key: values[0] for key, values in valid_options.items()})
+
+        assert info.default_config == default_config
+
+    @pytest.mark.parametrize(
+        "valid_options",
+        [
+            pytest.param(None, id="default"),
+            pytest.param(dict(option=("value",)), id="no_split"),
+        ],
+    )
+    def test_default_config_split_train(self, valid_options):
+        info = make_minimal_dataset_info(valid_options=valid_options)
+        assert info.default_config.split == "train"
+
+    def test_valid_options_split_but_no_train(self):
+        with pytest.raises(ValueError, match=re.escape("'train' has to be a valid argument for option 'split'")):
+            make_minimal_dataset_info(valid_options=dict(split=("test",)))
+
+    @pytest.mark.parametrize(
+        ("options", "expected_error_msg"),
+        [
+            pytest.param(dict(unknown_option=None), "Unknown option 'unknown_option'", id="unknown_option"),
+            pytest.param(dict(split="unknown_split"), "Invalid argument 'unknown_split'", id="invalid_argument"),
+        ],
+    )
+    def test_make_config_invalid_inputs(self, info, options, expected_error_msg):
+        with pytest.raises(ValueError, match=re.escape(expected_error_msg)):
+            info.make_config(**options)
+
+    def test_repr(self, info):
+        output = repr(info)
+
+        assert isinstance(output, str)
+        assert "DatasetInfo" in output
+        for key, value in info._valid_options.items():
+            assert f"{key}={str(value)[1:-1]}" in output
+
+    @pytest.mark.parametrize("optional_info", ("citation", "homepage", "license"))
+    def test_repr_optional_info(self, optional_info):
+        sentinel = "sentinel"
+        info = make_minimal_dataset_info(**{optional_info: sentinel})
+
+        assert f"{optional_info}={sentinel}" in repr(info)
+
+
+class TestDataset:
+    class DatasetMock(datasets.utils.Dataset):
+        def __init__(self, name="name", *, valid_options=None, resources=None):
+            self._name = name
+            self._valid_options = valid_options or dict(split=("train", "test"))
+
+            self.resources = unittest.mock.Mock(return_value=[]) if resources is None else lambda config: resources
+            self._make_datapipe = unittest.mock.Mock()
+
+        @property
+        def info(self):
+            return datasets.utils.DatasetInfo(
+                self._name,
+                type=datasets.utils.DatasetType.RAW,
+                categories=[],
+                valid_options=self._valid_options,
+            )
+
+        def resources(self, config):
```

Comment:
We can't. Methods are set on the class, so we actually need to overwrite the class attribute `resources`. Since we don't want the same return values for every instance, we need this dummy that is overwritten upon instantiation. Another solution would be to create the class at runtime, which is what I did before, but you didn't like that either :stuck_out_tongue: